iwyu.cc

```diff
@@ -4123,25 +4125,26 @@ class IwyuAstConsumer
     if (CanIgnoreCurrentASTNode())
       return true;
     // TypedefType::getDecl() returns the place where the typedef is defined.
-    if (CanForwardDeclareType(current_ast_node())) {
-      ReportDeclForwardDeclareUse(CurrentLoc(), type->getDecl());
-    } else {
-      ReportDeclUse(CurrentLoc(), type->getDecl());
-    }
+    ReportDeclUse(CurrentLoc(), type->getDecl());
+    if (!CanForwardDeclareType(current_ast_node()))
+      ReportTypeUse(CurrentLoc(), type);
     return Base::VisitTypedefType(type);
   }
 
   bool VisitUsingType(clang::UsingType* type) {
     if (CanIgnoreCurrentASTNode())
       return true;
 
-    // UsingType is similar to TypedefType, so treat it the same.
     if (CanForwardDeclareType(current_ast_node())) {
       ReportDeclForwardDeclareUse(CurrentLoc(), type->getFoundDecl());
     } else {
       ReportDeclUse(CurrentLoc(), type->getFoundDecl());
     }
 
+    const Type* underlying_type = type->getUnderlyingType().getTypePtr();
+    if (isa<TypedefType>(underlying_type))
```

Comment:
The special handling of `TypedefType` here is a little surprising. Don't you want to take `CanForwardDeclareType` into account here as well?

Reply:
This seems like a bug, thanks! Firstly, I wanted rather to call `VisitTypedefType` from here, but `UsingDecl` reporting should be changed in that case...

Reply:
It's probably dangerous to cross-call between `Visit` functions, you'd presumably have to extract some kind of helper to share code.